pkgs/development/compilers/haxe/default.nix

```diff
@@ -1,14 +1,38 @@
-{ lib, stdenv, fetchgit, coreutils, ocamlPackages, zlib, pcre, neko }:
-
-let inherit (ocamlPackages) ocaml camlp4; in
+{ lib, stdenv, fetchgit, coreutils, ocaml-ng, zlib, pcre, neko, mbedtls }:
 
 let
-  generic = { version, sha256, prePatch }:
+  ocamlDependencies = version:
+    if lib.versionAtLeast version "4.0"
+    then with ocaml-ng.ocamlPackages; [
+      ocaml
+      findlib
+      sedlex_2
+      xml-light
+      ptmap
+      camlp5
+      sha
+      dune_2
+      luv
+    ] ++ (if lib.versionAtLeast version "4.2" then [
+      ocaml_extlib_1_7_8
+    ] else [
+      ocaml_extlib
+    ])
+    else with ocaml-ng.ocamlPackages_4_05; [
+      ocaml
+      camlp4
+    ];
+
+  # extraPatch gets appended to the default prePatch. If prePatch is given,
+  # it overrides the default prePatch and extraPatch is irrelevant
```

Comment:
I'd keep it simple as in was.
Yes, some code was duplicated, but it did not require additional comment (with extraMentalLoad to a code reader)

Reply:
I mean in the original code and here https://github.com/NixOS/nixpkgs/issues/95638#issuecomment-776875554

Also, Haxe 4 is not yet well tested on NixOS and may require more patches.

Reply:
I used only NodeJs target

Reply:
Yeah, more testing would be great! I also suspect haxe 4.x needs one or two extra seds like the 3.x versions.

idk, there is only one special case so its worth it I'd say, especially if we'll remove 3.2 at some point.

Reply:
I meant all the `extraPatch` introduced for no profit, only `prePatch` was already enough

Reply:
Since we only really needed `extraPatch` in 3.4, doesn't seem that worth it actually. Removed in favor of a simpler model.